### PR TITLE
Master pbm sam arm

### PR DESCRIPTION
**Pull request by amoyaux**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/mrp_production.py [hunk 1]
```diff
@@ -163,6 +163,34 @@ def _get_default_location_dest_id(self):
...
+    delivery_count = fields.Integer(string='Delivery Orders', compute='_compute_picking_ids')
+
+    @api.multi
+    @api.depends('procurement_group_id')
+    def _compute_picking_ids(self):
+        for order in self:
+            warehouse = order.location_src_id.get_warehouse()
+            order.picking_ids = self.env['stock.picking'].search([
+                ('group_id', '=', order.procurement_group_id.id),
+                ('picking_type_id', 'in', [warehouse.manu_type_id.id, warehouse.pbm_type_id.id, warehouse.sam_type_id.id])])
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: is it really needed to filter also on the picking types?

File: addons/mrp/models/mrp_production.py [hunk 2]
```diff
@@ -163,6 +163,34 @@ def _get_default_location_dest_id(self):
     is_locked = fields.Boolean('Is Locked', default=True, copy=False)
     show_final_lots = fields.Boolean('Show Final Lots', compute='_compute_show_lots')
     production_location_id = fields.Many2one('stock.location', "Production Location", related='product_id.property_stock_production')
+    picking_ids = fields.Many2many('stock.picking', compute='_compute_picking_ids', string='Picking associated to this manufacturing order')
+    delivery_count = fields.Integer(string='Delivery Orders', compute='_compute_picking_ids')
+
+    @api.multi
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: this decorator isn't needed

File: addons/mrp/models/mrp_production.py [hunk 3]
```diff
@@ -163,6 +163,34 @@ def _get_default_location_dest_id(self):
...
+            order.delivery_count = len(order.picking_ids)
+
+    @api.multi
+    def action_view_mo_delivery(self):
+        '''
+        This function returns an action that display picking related to
+        manufacturing order orders. It can either be a in a list or in a form
+        view, if there is only one picking to show.
+        '''
+        action = self.env.ref('stock.action_picking_tree_all').read()[0]
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: there should be an ensure_one here

File: addons/mrp/models/stock_rule.py [hunk 4]
```diff
@@ -66,7 +66,7 @@ def _prepare_mo_vals(self, product_id, product_qty, product_uom, location_id, na
             'product_id': product_id.id,
             'product_qty': product_qty,
             'product_uom_id': product_uom.id,
-            'location_src_id': self.location_src_id.id or location_id.id,
+            'location_src_id': self.location_src_id.id or self.picking_type_id.default_location_src_id.id,
```
Comment (sle-odoo) [inline addons/mrp/models/stock_rule.py]: but why

File: addons/mrp/models/stock_warehouse.py [hunk 5]
```diff
@@ -12,103 +13,229 @@ class StockWarehouse(models.Model):
...
+        'Manufacture', default='mrp_one_step', required=True,
+        help="Produce : Move the raw materials to the production location\
+        directly and start the manufacturing process.\nPick / Produce : Unload\
+        the raw materials from the Stock to Input location first, and then\
+        transfer it to the Production location.")
+
+    manufacture_route_id = fields.Many2one('stock.location.route', 'Picking Before Manufacturing Route', ondelete='restrict')
+
+    wh_pbm_stock_loc_id = fields.Many2one('stock.location', 'Picking before Manufacturing Location')
+    wh_sam_stock_loc_id = fields.Many2one('stock.location', 'Stock after Manufacturing Location')
```
Comment (sle-odoo) [inline addons/mrp/models/stock_warehouse.py]: these names are weird

File: addons/stock/models/stock_warehouse.py [hunk 6]
```diff
@@ -78,6 +79,13 @@ def _compute_warehouse_count(self):
         for warehouse in self:
             warehouse.warehouse_count = self.env['stock.warehouse'].search_count([('id', 'not in', warehouse.ids)])
 
+    def _compute_show_resupply(self, show_resupply=False):
+        for warehouse in self:
+            if not show_resupply:
+                warehouse.show_resupply = not self.user_has_groups("stock.group_stock_multi_warehouses")
+            else:
+                warehouse.show_resupply = True
```
Comment (sle-odoo) [inline addons/stock/models/stock_warehouse.py]: we may want to have something cleaner here, let's think about it

File: addons/mrp/models/stock_warehouse.py [hunk 7]
```diff
@@ -12,103 +13,226 @@ class StockWarehouse(models.Model):
...
+        ('mrp_one_step', 'Manufacture (1 step)'),
+        ('pbm', 'Pick components and then manufacture (2 steps)'),
+        ('pbm_sam', 'Pick components, manufacture and then store products (3 steps)')],
+        'Manufacture', default='mrp_one_step', required=True,
+        help="Produce : Move the raw materials to the production location\
+        directly and start the manufacturing process.\nPick / Produce : Unload\
+        the raw materials from the Stock to Input location first, and then\
+        transfer it to the Production location.")
+
+    manufacture_route_id = fields.Many2one('stock.location.route', 'Picking Before Manufacturing Route', ondelete='restrict')
```
Comment (pimodoo) [inline addons/mrp/models/stock_warehouse.py]: rename in pbm_route_id

File: addons/mrp/models/stock_warehouse.py [hunk 8]
```diff
@@ -12,103 +13,226 @@ class StockWarehouse(models.Model):
...
+
+    pbm_loc_id = fields.Many2one('stock.location', 'Picking before Manufacturing Location')
+    sam_loc_id = fields.Many2one('stock.location', 'Stock after Manufacturing Location')
+
```
Comment (pimodoo) [inline addons/mrp/models/stock_warehouse.py]: add pbm_loc_id and sam_loc_id in technical warehouse view

File: addons/mrp/models/stock_warehouse.py [hunk 9]
```diff
@@ -12,103 +13,226 @@ class StockWarehouse(models.Model):
...
+                    'location_id': self.pbm_loc_id.id,
+                    'location_src_id': self.lot_stock_id.id,
+                    'picking_type_id': self.pbm_type_id.id
+                },
+                'update_values': {
+                    'active': self.manufacture_steps != 'mrp_one_step',
+                }
+            },
+            # The purpose to move sam rule in the manufacture route instead of
+            # manufacture_route_id is to avoid conflict with receipt in multiple
```
Comment (pimodoo) [inline addons/mrp/models/stock_warehouse.py]: rename also manufacture_route_id in this comment

File: addons/stock/models/stock_warehouse.py [hunk 10]
```diff
@@ -280,70 +385,35 @@ def _find_existing_rule_or_create(self, rules_list):
...
-        Rule = self.env['stock.rule']
-        for warehouse in self:
-            routings = routes_data[warehouse.id][warehouse.delivery_steps]
-            if warehouse.mto_pull_id:
-                mto_pull = warehouse.mto_pull_id
-                mto_pull.write(warehouse._get_mto_rules_values(routings)[0])
-            else:
-                mto_pull = Rule.create(warehouse._get_mto_rules_values(routings)[0])
-        return mto_pull
+    def _get_sublocations_values(self, vals):
```
Comment (pimodoo) [inline addons/stock/models/stock_warehouse.py]: maybe rename in _get_locations_values

File: addons/mrp_byproduct/wizard/change_production_qty.py [hunk 11]
```diff
@@ -20,4 +20,7 @@ def _update_product_to_produce(self, prod, qty):
                 qty1 *= product_uom_factor / prod.bom_id.product_qty
                 move[0].write({'product_uom_qty': qty1})
             else:
-                prod._create_byproduct_move(sub_product_line)
\ No newline at end of file
+                move = prod._create_byproduct_move(sub_product_line)
+            # TODO adjust qty
```
Comment (sle-odoo) [inline addons/mrp_byproduct/wizard/change_production_qty.py]: do it!

**Pull request merged**